Corrections in protocol, header/id and temperature

mrWheel opened the pull request:
https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700

This is my first pull-request! So please check and double-check ;-)

mrWheel commented:
There is a description of the alecto_ws1700 protocol.
The protocol exists of:
0101 11110011 0 0 00 000110010110 01010111

Each (group) of numbers has a specific meaning:

Header 0 till 3
ID: 4 till 11
Battery: 12
TX mode: 13
Channel: 14 till 15
Temperature: 16 till 27
Humidity: 28 till 35

The Header must always be 0b0101 (5)
If the first bit of Temperature = '1', it's a negative number

pilino1234 commented on libs/pilight/protocols/433.92/alecto_ws1700.c:
```diff
@@ -75,13 +75,17 @@ static void parseCode(void) {
-	if(temperature > 511) {
-		temperature -= 1023;
+	if (binary[16]) {
+	    temperature = temperature - 4096;
```
You can use the -= operator here

mrWheel replied:
Hi pilino1234,

Yes! you can. Normaly I would, just not paying attention.

As I’m quit new to the GitHub thingy: Can I just change the source and re-request for a "push-request”? Will the first one then disappear?

Please advice…!

Regards,

Willem Aandewiel
Almere

gsm: 0653 - 499 551
http: www.aandewiel.nl

---
:Core Wars: /n./  A game between `assembler' programs in a
   simulated machine, where the objective is to kill your opponent's
   program by overwriting it.  Popularized by A. K. Dewdney's column
   in "Scientific American" magazine, this was actually devised
   by Victor Vyssotsky, Robert Morris Sr., and Dennis Ritchie in the
   early 1960s (their original game was called `Darwin' and ran on a
   PDP-1 at Bell Labs).  See {core}.

> On 15 jun. 2016, at 13:33, pilino1234 <notifications@github.com> wrote:
> 
> In libs/pilight/protocols/433.92/alecto_ws1700.c <https://github.com/pilight/pilight/pull/287#discussion_r67145031>:
> 
> >  	humidity = (double)binToDecRev(binary, 28, 35);
> >  
> > -	if(temperature > 511) {
> > -		temperature -= 1023;
> > +	if (binary[16]) {
> > +	    temperature = temperature - 4096;
> You can use the -= operator here
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287/files/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887#r67145031>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6ExadIEE3sPTIcxA8Aut0mUpJUu8zks5qL-LzgaJpZM4I2DgO>.
>

pilino1234 replied:
Are you using Github to edit the code, or are you using git?

If you're using git you can make a new commit, and the follow these instructions to "squash" this current commit with the new one for the -= operator:
http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

Then you run
```git push origin HEAD -f```
to force push your change to the same branch. GitHub will automatically update the pull request.

Whatever you do, just please don't close this PR and open a new one :wink:

mrWheel replied:
Pffff…

Have to read that over (and over again..).

I have cloned pilight with GitHub to my own repository on my local machine.

From that I have cloned (from my local machine) the develop branche with “git clone" to my RPi.
There I have made the changes to the alecto_ws1700.c file, committed it with “git commit -a” and then pushed the changes to my local machine.
With GitHub I than reguested a “push-request”…

As I said, quit new to this Git(Hub) thingy...

I will give it a try in a bit.

Regards,

Willem

gsm: 0653 - 499 551
http: www.aandewiel.nl

---
:gobble: /vt./  1. To consume, usu. used with `up'.  "The
   output spy gobbles characters out of a {tty} output buffer."
   2. To obtain, usu. used with `down'.  "I guess I'll gobble down
   a copy of the documentation tomorrow."  See also {snarf}.

> On 15 jun. 2016, at 13:44, pilino1234 <notifications@github.com> wrote:
> 
> In libs/pilight/protocols/433.92/alecto_ws1700.c <https://github.com/pilight/pilight/pull/287#discussion_r67146322>:
> 
> >  	humidity = (double)binToDecRev(binary, 28, 35);
> >  
> > -	if(temperature > 511) {
> > -		temperature -= 1023;
> > +	if (binary[16]) {
> > +	    temperature = temperature - 4096;
> Are you using Github to edit the code, or are you using git?
> 
> If you're using git you can make a new commit, and the follow these instructions to "squash" this current commit with the new one for the -= operator:
> http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit <http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit>
> Then you run
> git push origin HEAD -f
> to force push your change to the same branch. GitHub will automatically update the pull request.
> 
> Whatever you do, just please don't close this PR and open a new one 😉
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287/files/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887#r67146322>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6E_m_c6Ud40O7Pt9yrYyTS66B6yM7ks5qL-W1gaJpZM4I2DgO>.
>

wo-rasp commented:
Hi Williem,

Try to stay away from "git commit -a", as it will commit more stuff then you may want to be pushed.

The best procedure is:
1. Use "git status" to get informed on what got changed. It will also tell you about inconsistencies and for most cases a suggestion on what to do next.
2. In your case it should only show "libs/pilight/protocols/433gpio/alecto_ws1700.c"
3. Use "git add libs/pilight/protocols/433gpio/alecto_ws1700.c" to add the file to the ones to be commited.
4. Use "git commit" and add your reason in the text editor.
5. Use "git push origin development" to upload it to your github account.
6. On github issue the pull request.

Now you have changes to your original commit.
In step 4, do not use "git commit" but use instead "git commit --amend"
That will not create a new log entry, but amend your 2nd (3rd, 4th, ...) commit to the original one.
Do Step 5, there is no need for Step 6, as the pull request is remaining valid.

A general word on using the development branch for your development, technically you can do so, but in practive, it will cause conflicts, in particular if the maintainer of the project makes changes to your commit, and you want to merge back the results with your local copy.

It is better if you use a separate branch for each intended modification. It is very easy to do, Checkout the branch you want to add modifications to, e.q. development.
Now use the command "git checkout -b branchname" to create a new branch. This branch has all stuff from development, plus your changes. Once your changes are merged into development, you can delete this branch.

Wolfgang

mrWheel commented:
Good afternoon Wolfgang,

That is a clear handout .. but I already have committed my change to my local (and GitHub) repository.
I now have this message on the screen:

This branch is 2 commits ahead of pilight:development.

If I click on Pull-Request I see there are two commits:


Do I have to press any button in this screen or is the request already merged and waiting for approval?

It say’s [Unified] but I can select the two commits to do a combined request… Pfff.
I’m probably getting to old (just turned 61) for this high tech stuff ;-)

Thanks again for your help and patience!

Regards,

Willem

gsm: 0653 - 499 551
http: www.aandewiel.nl

---
I am looking for a honest man.
		-- Diogenes the Cynic

> On 15 jun. 2016, at 14:43, wo-rasp <notifications@github.com> wrote:
> 
> Hi Williem,
> 
> Try to stay away from "git commit -a", as it will commit more stuff then you may want to be pushed.
> 
> The best procedure is:
> 1. Use "git status" to get informed on what got changed. It will also tell you about inconsistencies and for most cases a suggestion on what to do next.
> 2. In your case it should only show "libs/pilight/protocols/433gpio/alecto_ws1700.c"
> 3. Use "git add libs/pilight/protocols/433gpio/alecto_ws1700.c" to add the file to the ones to be commited.
> 4. Use "git commit" and add your reason in the text editor.
> 5. Use "git push origin development" to upload it to your github account.
> 6. On github issue the pull request.
> 
> Now you have changes to your original commit.
> In step 4, do not use "git commit" but use instead "git commit --amend"
> That will not create a new log entry, but amend your 2nd (3rd, 4th, ...) commit to the original one.
> Do Step 5, there is no need for Step 6, as the pull request is remaining valid.
> 
> A general word on using the development branch for your development, technically you can do so, but in practive, it will cause conflicts, in particular if the maintainer of the project makes changes to your commit, and you want to merge back the results with your local copy.
> 
> It is better if you use a separate branch for each intended modification. It is very easy to do, Checkout the branch you want to add modifications to, e.q. development.
> Now use the command "git checkout -b branchname" to create a new branch. This branch has all stuff from development, plus your changes. Once your changes are merged into development, you can delete this branch.
> 
> Wolfgang
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287#issuecomment-226175759>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6E1re6Dr-lHiMymx6ZQggxtyToTkhks5qL_OHgaJpZM4I2DgO>.
>

mrWheel commented:
If I click on Pull-Request I see there are two commits:

![screen shot 2016-06-15 at 14 49 03](https://cloud.githubusercontent.com/assets/5585427/16080835/f75b9948-330a-11e6-8b77-4d57a6680aa2.png)

wo-rasp commented:
Okay, you did two commits, and you can see them.

Let us start with the local repository and let git guide you.

Four prerequisited are important to understand:
A) In principle git is working with commits only:
- the current one (The HEAD), the previous one and the next one you create
B) A branch points to a single commit
- nothing more and nothing less.
C) You can fork (duplicate) all files of a given commit
- normally you give this fork a branch name
D) you can change the predecessor of a commit
- all successors will inherit changes from the predecessor

To have a look at what's going on with your sequence of commits "git log" is your friend and you do need to memorize the following four options:
--all
that will enforce to show the sequence of commits of your current branch.
--oneline
that will the amount of information to the sha code and your commit message
--graph
that will print the sequence of commits
--decorate
- this will add the branch names

Just get familiar with the various options.


To combine your two commits you do need to ensure that your HEAD is at the most recent commit "git log -all --oneline HEAD -2" will tell you where you are. The option -2 means the number of commits to be logged, so you should see two lines starting with f7409c and 22bebd7.

Once you are sure that you have the correct number of commits you want to combine, there are multiple options to change your history, i personally prefer the rebase option:
- "git rebase --i HEAD~2" and your favorite editor opens up.

Leave in the first line the command pick unchanged and change in the 2nd line the command pick into squash, (have you noted that the sequence of commits in the file is just in reverse order ?). Thus the commit 22bebd7 is "picked" as base and the files from commit f74c09c are squashed (merged) into it and your job is done.

You have done the changes in your local repository only.
When you use git log --oneline --decorate--graph --all you will notice that HEAD and development are in one line, but origin/development is on another line.

You do need to push your local changes to github ("git push origin development") and when you use got log again, HEAD, development origin/development) are all on one line.

mrWheel commented:
Dear Wolfgang,

I’m afraid I broke something….

============= console-log ====================

~/pilight# git log HEAD -2
commit 22bebd7a43065c65f2fe81bbe36a0f38b84a6d74
Author: TheWheel <Willem@Aandewiel.nl>
Date:   Wed Jun 15 14:30:05 2016 +0200

    Change protocol Head, ID and Temperature conform specsheet.
    https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17682#pid17682

commit f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887
Author: TheWheel <Willem@Aandewiel.nl>
Date:   Wed Jun 15 08:52:48 2016 +0200

    Corrections in protocol, header/id and temperature
    https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700

~/pilight# git rebase -i HEAD~2
error: could not apply 22bebd7... Change protocol Head, ID and Temperature conform specsheet.

When you have resolved this problem run "git rebase --continue".
If you would prefer to skip this patch, instead run "git rebase --skip".
To check out the original branch and stop rebasing run "git rebase --abort".
Could not apply 22bebd7... Change protocol Head, ID and Temperature conform specsheet. 

~/pilight# git log --all --oneline HEAD -2
22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
f74c09c Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-se

~/pilight# git rebase -i HEAD~2

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
	git rebase (--continue | --abort | --skip)
If that is not the case, please
	rm -fr /root/pilight/.git/rebase-merge
and run me again.  I am stopping in case you still have something
valuable there.

~/pilight# rm -fr /root/pilight/.git/rebase-merge

~/pilight# git rebase -i HEAD~2
libs/pilight/protocols/433.92/alecto_ws1700.c: needs merge
Cannot rebase: You have unstaged changes.
Additionally, your index contains uncommitted changes.
Please commit or stash them.

~/pilight# git commit -a
[detached HEAD 1444524] Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17682#pid17682
 1 file changed, 5 insertions(+)

~/pilight# git reset HEAD libs/pilight/protocols/433.92/alecto_ws1700.c

~/pilight# git commit -a

~/pilight# git add libs/pilight/protocols/433.92/alecto_ws1700.c

~/pilight# git commit
# Not currently on any branch.

~/pilight# git log --all --oneline HEAD -2
1444524 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
root@pilight7:~/pilight# git rebase -i HEAD~2
Successfully rebased and updated detached HEAD.

~/pilight# git commit --amend
[detached HEAD a87b0ac] Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378
 Author: wo-rasp <wo-rasp@users.noreply.github.com>
 1 file changed, 3 insertions(+), 3 deletions(-)

~/pilight# 

I give up (for now)

Thanks for all your help,
Regards

Willem

gsm: 0653 - 499 551
http: www.aandewiel.nl

---
America, how can I write a holy litany in your silly mood?
		-- Allen Ginsberg

> On 15 jun. 2016, at 17:22, wo-rasp <notifications@github.com> wrote:
> 
> Okay, you did two commits, and you can see them.
> 
> Let us start with the local repository and let git guide you.
> 
> Four prerequisited are important to understand:
> A) In principle git is working with commits only:
> 
> the current one (The HEAD), the previous one and the next one you create B) A branch points to a single commit
> nothing more and nothing less. C) You can fork (duplicate) all files of a given commit
> normally you give this fork a branch name D) you can change the predecessor of a commit
> all successors will inherit changes from the predecessor
> To have a look at what's going on with your sequence of commits "git log" is your friend and you do need to memorize the following four options:
> --all
> that will enforce to show the sequence of commits of your current branch.
> --oneline
> that will the amount of information to the sha code and your commit message
> --graph
> that will print the sequence of commits
> --decorate
> 
> this will add the branch names
> Just get familiar with the various options.
> 
> To combine your two commits you do need to ensure that your HEAD is at the most recent commit "git log -all --oneline HEAD -2" will tell you where you are. The option -2 means the number of commits to be logged, so you should see two lines starting with f7409c and 22bebd7 <https://github.com/pilight/pilight/commit/22bebd7a43065c65f2fe81bbe36a0f38b84a6d74>.
> 
> Once you are sure that you have the correct number of commits you want to combine, there are multiple options to change your history, i personally prefer the rebase option:
> 
> "git rebase --i HEAD~2" and your favorite editor opens up.
> Leave in the first line the command pick unchanged and change in the 2nd line the command pick into squash, (have you noted that the sequence of commits in the file is just in reverse order ?). Thus the commit 22bebd7 <https://github.com/pilight/pilight/commit/22bebd7a43065c65f2fe81bbe36a0f38b84a6d74> is "picked" as base and the files from commit f74c09c <https://github.com/pilight/pilight/commit/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887> are squashed (merged) into it and your job is done.
> 
> You have done the changes in your local repository only.
> When you use git log --oneline --decorate--graph --all you will notice that HEAD and development are in one line, but origin/development is on another line.
> 
> You do need to push your local changes to github ("git push origin development") and when you use got log again, HEAD, development origin/development) are all on one line.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287#issuecomment-226221621>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6E2HTafeQrpNxBHFS581BdOsdgLNXks5qMBiegaJpZM4I2DgO>.
>

wo-rasp commented:
It is very difficuklt to break something in GIT.
Let me have a look at it.

mrWheel commented:
Well,
To quote Douglas Adams:
A common mistake that people make when trying to design something completely foolproof is to underestimate the ingenuity of complete fools.

Regards,

Willem 

gsm: 0653 - 499 551
www: http://willem.aandewiel.nl/


---
Thank God I'm Atheist

Verstuurd vanaf mijn iPad

> Op 16 jun. 2016 om 08:02 heeft wo-rasp <notifications@github.com> het volgende geschreven:
> 
> It is very difficuklt to break something in GIT.
> Let me have a look at it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

wo-rasp commented:
Okay, your github repository is okay.
I suggest to discuss more details in the pilight forum, below are the steps to recover quickly:
clone your github repository.
git checkout development
git rebase -i HEAD~2
In the editor windows edit the 2nd line (replace pick with squash)::
pick f74c09c Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws17$
squash 22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alec$
Save the results
The editor opens again, save that file and you have your new commit to be committed and pushed to your github repository.

mrWheel commented:
Hi Wolfgang,

Did exactly what you wrote (except for the error with “git push origin development"
If I now look at pilight/pilight I see four(!!!!) commits instead of the two I had before and three more than the “One” I expected!!!

f74c09c, 22bebd7, 3939b77 and b40e91a

See Attachment.

I wonder if this is it, if it’s still wrong and what I can do next.

=========== console log ======================================
~/pilight# git checkout development
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  a87b0ac Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch new_branch_name a87b0ac5fbbe77b76b32dbb61db73a13acfe79d6

Switched to branch 'development'

~/pilight# git log --all --oneline HEAD -2
22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
f74c09c Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-se

~/pilight# git rebase -i HEAD~2
[detached HEAD 3939b77] Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700
 1 file changed, 9 insertions(+), 5 deletions(-)
Successfully rebased and updated refs/heads/development.

~/pilight# git push origin development
Username for 'https://github.com': Willem@Aandewiel.nl
Password for 'https://Willem@Aandewiel.nl@github.com': 
To https://github.com/mrWheel/pilight.git
 ! [rejected]        development -> development (non-fast-forward)
error: failed to push some refs to 'https://github.com/mrWheel/pilight.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Merge the remote changes (e.g. 'git pull')
hint: before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

~/pilight# git pull
Merge made by the 'recursive' strategy.
root@pilight7:~/pilight# git log --all --oneline HEAD -2
b40e91a Merge branch 'development' of https://github.com/mrWheel/pilight into development
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu

~/pilight# git push origin development
Username for 'https://github.com': Willem@Aandewiel.nl
Password for 'https://Willem@Aandewiel.nl@github.com': 
Counting objects: 2, done.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 579 bytes, done.
Total 2 (delta 0), reused 0 (delta 0)
To https://github.com/mrWheel/pilight.git
   22bebd7..b40e91a  development -> development

~/pilight# 
=========== console log ======================================

I must say: I’m impressed by git/github but I have no idear what I’m doing.
It is not due to the help you give me, that’s beyond what can be expected and makes me feel I’m wasting your time.
I must start investigating the git(hub) more thoroughly because I don’t understand most of the terminology. Have used SVN a lot but only as a single developer. No branches or other stuff, just “svn add”, “svn commit” and “svn update”…
 
Can you confirm I’ve done all there is and the changes will be incorporated in the development branche?


I don’t know how to follow up on your suggestion “to discuss more details in the pilight forum”...

Thanks again for your time and help! Where would I be without you.

Regards,

Willem

gsm: 0653 - 499 551
http: www.aandewiel.nl

---
We don't know who it was that discovered water, but we're pretty sure
that it wasn't a fish.
	-- Marshall McLuhan

> On 16 jun. 2016, at 09:39, wo-rasp <notifications@github.com> wrote:
> 
> Okay, your github repository is okay.
> I suggest to discuss more details in the pilight forum, below are the steps to recover quickly:
> clone your github repository.
> git checkout development
> git rebase -i HEAD~2
> In the editor windows edit the 2nd line (replace pick with squash)::
> pick f74c09c <https://github.com/pilight/pilight/commit/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887> Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws17$ <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws17%24>
> squash 22bebd7 <https://github.com/pilight/pilight/commit/22bebd7a43065c65f2fe81bbe36a0f38b84a6d74> Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alec$ <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alec%24>
> Save the results
> The editor opens again, save that file and you have your new commit to be committed and pushed to your github repository.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287#issuecomment-226410840>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6E07eelU457Gq-ltXXnreSL3jVbHeks5qMP2mgaJpZM4I2DgO>.
>

pilino1234 commented:
You need to use ```git push origin development -f``` after rebasing. Note the "-f" option to force the push - this is necessary because the commit history of the repository on your computer looks different to the one on GitHub.

mrWheel commented:
=========================================
~/pilight# git push origin development -f
Username for 'https://github.com': Willem@Aandewiel.nl
Password for 'https://Willem@Aandewiel.nl@github.com': 
Everything up-to-date
~/pilight# 
=========================================

I see no changes in the preview

pilino1234 commented:
Hmm, now you need to do the rebasing again (git rebase -i ........), but instead of running ```git pull``` and ```git push origin development```, you just run ```git push origin development -f```. That should do it.

mrWheel commented:
~/pilight# git rebase -i HEAD~2
error: could not apply f74c09c... Corrections in protocol, header/id and temperature

When you have resolved this problem run "git rebase --continue".
If you would prefer to skip this patch, instead run "git rebase --skip".
To check out the original branch and stop rebasing run "git rebase --abort".

Could not apply f74c09c... Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700

~/pilight# git log --all --oneline HEAD -2
b40e91a Merge branch 'development' of https://github.com/mrWheel/pilight into development
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu

~/pilight# git log --all --oneline HEAD -3
b40e91a Merge branch 'development' of https://github.com/mrWheel/pilight into development
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
~/pilight# git rebase -i HEAD~3

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
	git rebase (--continue | --abort | --skip)
If that is not the case, please
	rm -fr /root/pilight/.git/rebase-merge
and run me again.  I am stopping in case you still have something
valuable there.

~/pilight# rm -fr /root/pilight/.git/rebase-merge

~/pilight# git rebase -i HEAD~3
libs/pilight/protocols/433.92/alecto_ws1700.c: needs merge
Cannot rebase: You have unstaged changes.
Additionally, your index contains uncommitted changes.
Please commit or stash them.

~/pilight# git commit -a
[detached HEAD fa2839c] Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700
 1 file changed, 13 insertions(+), 5 deletions(-)

~/pilight# git log --all --oneline HEAD -4
fa2839c Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
b40e91a Merge branch 'development' of https://github.com/mrWheel/pilight into development
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu
22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-hu

~/pilight# git rebase -i HEAD~4
[detached HEAD da569ab] Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700 Bugfixing tcm protocol: structure tfa-> chenged to tcm-> for both battery objects
 **Author: wo-rasp <wo-rasp@users.noreply.github.com**
 4 files changed, 21 insertions(+), 13 deletions(-)
Successfully rebased and updated detached HEAD.

~/pilight# git push origin development -f
Username for 'https://github.com': Willem@Aandewiel.nl
Password for 'https://Willem@Aandewiel.nl@github.com': 
Everything up-to-date

Eh.... but no changes in github, still 4 files

wo-rasp commented:
Hi Willem,

Welcome to the world of git.

Your problem occured when you issued the git pull command.
But i will check into it in more details tonight
I will create an GIT topic on the forum, as i think we should not discuss the handling of github here, but keep the github comments for the original purpose to discuss technical issues about the commit itself.

Wolfgang

wo-rasp commented:
Hi Willem,

To me it is not fully clear when and how you created commit 3939b77. 
It was not part of the guthub account, when i cloned your repository Thu 16th in the morning.

But that is in my opinion the commit you wanted to create.
The commit was created on Wednesday, the 15th at 8:52:48
at the same time you created the commit f74c09c.
(You incorporated the comment from pilino: temperature -= 4096;)
The remaining task is now to remove the unnecessary commits.
As a 1st step checkout your development branch (git checkout development).

Use git log to determine how many commits you want to remove.
Starting with git log --all --oneline --decorate --graph -n 5 you see commits that you want to keep, most notably 3939b77 and you see commit a2719cf which is (at the time of this writing the HEAD of the development branch in the uostream repository, and it is the commit your changes are based on.
git log --all --oneline --decorate --graph -n 1 shows commit b40e91a only, so we now do know the number of commits you want to remove starting from the HEAD.
ENSURE THAT THE HEAD IS AT b40e91a !!!
You use git reset to get rid of unwanted commits.

root@pi_65:/home/pi/willem_1/pilight# git reset HEAD~1 --soft
root@pi_65:/home/pi/willem_1/pilight# git log --oneline -n 3
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700
a2719cf Merge pull request #285 from wo-rasp/dev_optional
627c98f Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378

Mission accomplished. 

Wolfgang

mrWheel commented:
Good morning Wolfgang,

Do you still need me to follow these steps as it "looks" like you already did them...
I don't want to mess up more than I already did!

Regards,

Willem 

gsm: 0653 - 499 551
www: http://willem.aandewiel.nl/


---
Thank God I'm Atheist

Verstuurd vanaf mijn iPad

> Op 17 jun. 2016 om 01:36 heeft wo-rasp <notifications@github.com> het volgende geschreven:
> 
> Hi Willem,
> 
> To me it is not fully clear when and how you created commit 3939b77. 
> It was not part of the guthub account, when i cloned your repository Thu 16th in the morning.
> 
> But that is in my opinion the commit you wanted to create.
> The commit was created on Wednesday, the 15th at 8:52:48
> at the same time you created the commit f74c09c.
> (You incorporated the comment from pilino: temperature -= 4096;)
> The remaining task is now to remove the unnecessary commits.
> As a 1st step checkout your development branch (git checkout development).
> 
> Use git log to determine how many commits you want to remove.
> Starting with git log --all --oneline --decorate --graph -n 5 you see commits that you want to keep, most notably 3939b77 and you see commit a2719cf which is (at the time of this writing the HEAD of the development branch in the uostream repository, and it is the commit your changes are based on.
> git log --all --oneline --decorate --graph -n 1 shows commit b40e91a only, so we now do know the number of commits you want to remove starting from the HEAD.
> ENSURE THAT THE HEAD IS AT b40e91a !!!
> You use git reset to get rid of unwanted commits.
> 
> root@pi_65:/home/pi/willem_1/pilight# git reset HEAD~1 --soft
> root@pi_65:/home/pi/willem_1/pilight# git log --oneline -n 3
> 3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700
> a2719cf Merge pull request #285 from wo-rasp/dev_optional
> 627c98f Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378
> 
> Mission accomplished.
> 
> Wolfgang
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mrWheel commented:
Hi Wolfgang

I followed your instructions:
================ terminal output ===============
~/pilight# git checkout development
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  da569ab Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700 Bugfixing tcm protocol: structure tfa-> chenged to tcm-> for both battery objects

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch new_branch_name da569ab3ab8169ef62a22b01bca185150581cb46

Switched to branch 'development'

~/pilight# git log --all --oneline --decorate --graph -n 5
*   b40e91a (HEAD, origin/development, development) Merge branch 'development' of https://github.co
|\  
| * 22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/T
| * f74c09c Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Ful
* | 3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/T
|/  
*   a2719cf Merge pull request #285 from wo-rasp/dev_optional
|\  
root@pilight7:~/pilight# git log --all --oneline --decorate --graph -n 1
*   b40e91a (HEAD, origin/development, development) Merge branch 'development' of https://github.co
|\  

~/pilight# git reset HEAD~1 --soft

~/pilight# git log --oneline -n 3
3939b77 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Threa
a2719cf Merge pull request #285 from wo-rasp/dev_optional
627c98f Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon

~/pilight# 
================ terminal output ===============
   
Should I not end with "git push origin development -f”?

Regards,

Willem Aandewiel

gsm: 0653 - 499 551
web: http://willem.aandewiel.nl/
—

A common mistake that people make when trying to design something completely foolproof is to underestimate the ingenuity of complete fools.
[Douglas Adams]

> On 17 Jun 2016, at 01:36, wo-rasp <notifications@github.com> wrote:
> 
> Hi Willem,
> 
> To me it is not fully clear when and how you created commit 3939b77 <https://github.com/pilight/pilight/commit/3939b775919b8b4f1ca45e01dbf3314aa9481316>. 
> It was not part of the guthub account, when i cloned your repository Thu 16th in the morning.
> 
> But that is in my opinion the commit you wanted to create.
> The commit was created on Wednesday, the 15th at 8:52:48
> at the same time you created the commit f74c09c <https://github.com/pilight/pilight/commit/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887>.
> (You incorporated the comment from pilino: temperature -= 4096;)
> The remaining task is now to remove the unnecessary commits.
> As a 1st step checkout your development branch (git checkout development).
> 
> Use git log to determine how many commits you want to remove.
> Starting with git log --all --oneline --decorate --graph -n 5 you see commits that you want to keep, most notably 3939b77 <https://github.com/pilight/pilight/commit/3939b775919b8b4f1ca45e01dbf3314aa9481316> and you see commit a2719cf <https://github.com/pilight/pilight/commit/a2719cf492230146edf276baaf056121f3b54533> which is (at the time of this writing the HEAD of the development branch in the uostream repository, and it is the commit your changes are based on.
> git log --all --oneline --decorate --graph -n 1 shows commit b40e91a <https://github.com/pilight/pilight/commit/b40e91af3727cf6bf18ba2d874cd6824cb74d457> only, so we now do know the number of commits you want to remove starting from the HEAD.
> ENSURE THAT THE HEAD IS AT b40e91a <https://github.com/pilight/pilight/commit/b40e91af3727cf6bf18ba2d874cd6824cb74d457> !!!
> You use git reset to get rid of unwanted commits.
> 
> root@pi_65:/home/pi/willem_1/pilight# git reset HEAD~1 --soft
> root@pi_65:/home/pi/willem_1/pilight# git log --oneline -n 3
> 3939b77 <https://github.com/pilight/pilight/commit/3939b775919b8b4f1ca45e01dbf3314aa9481316> Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700 <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700>
> a2719cf <https://github.com/pilight/pilight/commit/a2719cf492230146edf276baaf056121f3b54533> Merge pull request #285 <https://github.com/pilight/pilight/pull/285> from wo-rasp/dev_optional
> 627c98f <https://github.com/pilight/pilight/commit/627c98f4ab0997cae69d10c6a6ea96c4deb8b42b> Update json elements with conftype DEVICES_OPTIONAL https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378 <https://forum.pilight.org/Thread-Oregon-THGR122NX?pid=18378#pid18378>
> Mission accomplished.
> 
> Wolfgang
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287#issuecomment-226643863>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6E-2V9tY2RZJXwHgE5aL5rFIyXYwSks5qMd3kgaJpZM4I2DgO>.
>

wo-rasp commented:
Hi Willem,

Yes you need to push your results including th -f option.
The reason I asked you not to do so, is to have a backup available in case something goes wrong.
Until you are familiar with git please do yourself a favour and always use the --decorate --graph --all options when you use got log. If you want to avoid the typing, create an alias.
The --graph option does show you that all 3 commits are in a sequence (the asteriks symbol),
the --decorate option does show you were the HEAD is
the --all option does show you all commits not just the ones you are working with.

Doing so "git log --all --decorate --graph --oneline" will tell give the following brief overview:
- that your HEAD and your stageing area is 3 commits behind your github repositories (origin/development) - and that you therefore must use the -f option when you want to push your work to your github repository:

If you perform "git push origin development -f" (please note that there is a space between origin and development, not a /), commit 3939b77 will be the new head of origin/development and the predecessor is the commit you started to work with.

root@pi_65:/home/pi/willem_1/pilight# git log --all --oneline --decorate --graph -n 5
*   b40e91a (origin/development) Merge branch 'development' of https://github.com/mrWheel/pilight into development
|\
| * 22bebd7 Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17682#pid17682
| * f74c09c Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700
* | 3939b77 (HEAD, development) Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700
|/
*   a2719cf Merge pull request #285 from wo-rasp/dev_optional

Wolfgang

mrWheel commented:
Wolfgang,

Thanks again!
I think that after twenty tries it worked at once ;-)

I can not thank you enough!

I have placed a reply on your GIT post.

Deepest regards,

Willem Aandewiel

gsm: 0653 - 499 551
web: http://willem.aandewiel.nl/
—

A common mistake that people make when trying to design something completely foolproof is to underestimate the ingenuity of complete fools.
[Douglas Adams]

> On 18 Jun 2016, at 02:41, wo-rasp <notifications@github.com> wrote:
> 
> Hi Willem,
> 
> Yes you need to push your results including th -f option.
> The reason I asked you not to do so, is to have a backup available in case something goes wrong.
> Until you are familiar with git please do yourself a favour and always use the --decorate --graph --all options when you use got log. If you want to avoid the typing, create an alias.
> The --graph option does show you that all 3 commits are in a sequence (the asteriks symbol),
> the --decorate option does show you were the HEAD is
> the --all option does show you all commits not just the ones you are working with.
> 
> Doing so "git log --all --decorate --graph --oneline" will tell give the following brief overview:
> 
> that your HEAD and your stageing area is 3 commits behind your github repositories (origin/development) - and that you therefore must use the -f option when you want to push your work to your github repository:
> If you perform "git push origin development -f" (please note that there is a space between origin and development, not a /), commit 3939b77 <https://github.com/pilight/pilight/commit/3939b775919b8b4f1ca45e01dbf3314aa9481316> will be the new head of origin/development and the predecessor is the commit you started to work with.
> 
> root@pi_65:/home/pi/willem_1/pilight# git log --all --oneline --decorate --graph -n 5
> 
> b40e91a <https://github.com/pilight/pilight/commit/b40e91af3727cf6bf18ba2d874cd6824cb74d457> (origin/development) Merge branch 'development' of https://github.com/mrWheel/pilight <https://github.com/mrWheel/pilight> into development |\ | * 22bebd7 <https://github.com/pilight/pilight/commit/22bebd7a43065c65f2fe81bbe36a0f38b84a6d74> Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17682#pid17682 <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17682#pid17682> | * f74c09c <https://github.com/pilight/pilight/commit/f74c09ce59a0ffa3dfe12accbe3a1a1bf8580887> Corrections in protocol, header/id and temperature https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700 <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700?pid=17700#pid17700>
> | 3939b77 <https://github.com/pilight/pilight/commit/3939b775919b8b4f1ca45e01dbf3314aa9481316> (HEAD, development) Change protocol Head, ID and Temperature conform specsheet. https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700 <https://forum.pilight.org/Thread-Fully-Supported-No-brand-temp-humidity-sensor-alecto-ws1700> |/
> a2719cf <https://github.com/pilight/pilight/commit/a2719cf492230146edf276baaf056121f3b54533> Merge pull request #285 <https://github.com/pilight/pilight/pull/285> from wo-rasp/dev_optional
> Wolfgang
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pilight/pilight/pull/287#issuecomment-226911165>, or mute the thread <https://github.com/notifications/unsubscribe/AFU6ExOGfSF00LkNkWnCA9AlxsadkORjks5qMz69gaJpZM4I2DgO>.
>

pilino1234 commented:
You should also bump the version number of the protocol to 2.4 ...

mrWheel commented:
Dare I try?

Hope you have time to write the "best practise on pilight git use".
Maybe I better wait for that.

Regards,

Willem 

gsm: 0653 - 499 551
www: http://willem.aandewiel.nl/


---
Thank God I'm Atheist

Verstuurd vanaf mijn iPad

> Op 19 jun. 2016 om 13:10 heeft pilino1234 <notifications@github.com> het volgende geschreven:
> 
> You should also bump the version number of the protocol to 2.4 ...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

wo-rasp commented:
Hmm. IMHO: No - there is no need to !!
Willem finished the job already last Wednesday with commit 3939b77.
I do not know how he ended up there, i assume it happened "unmonitored" as he was working with two different repositories (at least he said that he cloned the repo to one machine, from there he cloned the repo to his Pi, and it is in the dark what exactly happened, as the sequence of events and the time stamps are confusing as well).
The only topic that counts is that commit 3939b77 has all work integrated (i have provided as part of my git post the steps required to verify that the commit is indeed correct using "git diff")
He just needs to reset the index in his staging area (that is the purpose of the --soft option), and push the results as described.

On the other side, he should have used the "git commit --amend" after his first commit f74c99c.

pilino1234 commented:
Looks good! Will test soon.

pilino1234 commented:
Been running this over the past few days, it works well with the 4 sensors here.
